signer/migrations/0011__materialized_chains.sql

```diff
@@ -0,0 +1,115 @@
+-- This table holds the canonical chain information for Bitcoin and Stacks
+-- chains keyed by the bitcoin chain tip.
+CREATE TABLE sbtc_signer.canonical_chains (
+    run_id INT,
+    bitcoin_chain_tip BYTEA,
+    bitcoin_block_hash BYTEA,
+    bitcoin_block_height BIGINT,
+    stacks_block_hash BYTEA,
+    stacks_block_height BIGINT
+);
+
+-- Indexes to support common queries on the canonical chains table.
+CREATE INDEX ix_canonical_chains_run_id ON sbtc_signer.canonical_chains(run_id);
+CREATE INDEX ix_canonical_chains_bitcoin_chain_tip ON sbtc_signer.canonical_chains(bitcoin_chain_tip);
+CREATE INDEX ix_canonical_chains_bitcoin_block_hash ON sbtc_signer.canonical_chains(bitcoin_block_hash);
+CREATE INDEX ix_canonical_chains_bitcoin_block_height ON sbtc_signer.canonical_chains(bitcoin_block_height);
+CREATE INDEX ix_canonical_chains_stacks_block_hash ON sbtc_signer.canonical_chains(stacks_block_hash);
+CREATE INDEX ix_canonical_chains_stacks_block_height ON sbtc_signer.canonical_chains(stacks_block_height);
+
+-- New sequence which will be used to generate the run_id for each materialized
+-- view canonical chain.
+CREATE SEQUENCE sbtc_signer.canonical_chains_run_id_seq;
+
+-- Function to materialize the canonical chains for a given bitcoin chain tip
+-- and the maximum depth of the bitcoin blockchain to consider.
+--
+-- This function returns the number of rows written to the `canonical_chains`
+-- table upon success, and `-1` if rows already exist for the given chain tip.
+CREATE OR REPLACE FUNCTION sbtc_signer.materialize_canonical_chains(chain_tip BYTEA, max_depth INT)
+RETURNS INTEGER AS $$
+DECLARE
+    rows_written INTEGER;
+	existing_rows INTEGER;
+	new_run_id BIGINT;
+BEGIN
+	-- Get the next value for the run_id
+    new_run_id := nextval('sbtc_signer.canonical_chains_run_id_seq');
+
+	-- Check if rows exist with the given chain_tip
+    SELECT COUNT(*) INTO existing_rows
+    FROM sbtc_signer.canonical_chains
+    WHERE bitcoin_chain_tip = chain_tip;
+
+    -- If rows exist, return an error code
+    IF existing_rows > 0 THEN
+        RETURN -1; -- Error code indicating rows already exist
+    END IF;
```

Comment:
I'm not sure about this: we get new blocks from the block observer when it thinks we have a better block, but we use our own logic to determine the chain tip -- in particular wrt the height tie breaking; so I think it can happen that we call this function twice for the same chain tip. I'm moderately sure it doesn't matter for the bitcoin side (like, we just need to handle the error correctly outside), but it could matter for the stacks side since we may have new stacks blocks?
Also, do we need to lock/run this whole function in a transaction? Because it's fine for the coordinator, but the signers may start to query this for a bitcoin tip that is still being worked on.

Reply:
I had "overwriting" of the chain-tip in my first iteration but removed it, primarily to avoid that some rogue code accidentally calls it twice for the same chain-tip (we don't have any PK's/UK's here) -- which is why I chose to return an error code so that the calling code could detect this and choose to separately remove the chain tip and replace it. That's how I'm thinking anyway 🤷

Reply:
That makes sense 👍